File: c/knapsack/knapsack.c
#include "knapsack.h"

unsigned int maximum_value(unsigned int max_weight, item_t* items, size_t item_count) {
    if (item_count == 0 || max_weight == 0) return 0;

    item_t item = items[item_count - 1];
    unsigned int picked = 0, not_picked = 0;
    if (item.weight <= max_weight) {
        picked = maximum_value(max_weight - item.weight, items, item_count - 1) + item.value;
    }
    not_picked = maximum_value(max_weight, items, item_count - 1);

    if (picked > not_picked) return picked;
    return not_picked;
}

File: zig/knapsack/knapsack.zig
const std = @import("std");
const mem = std.mem;

pub const Item = struct {
    // This struct, as well as its fields and init method, needs to be implemented.
    weight: usize,
    value: usize,

    pub fn init(weight: usize, value: usize) Item {
        return Item{
            .weight = weight,
            .value = value,
        };
    }
};

pub fn maximumValue(allocator: mem.Allocator, maximumWeight: usize, items: []const Item) !usize {
    if (items.len == 0) return 0;

    const item = items[0];
    if (item.weight > maximumWeight) return 0;

    const pickValue = (maximumValue(allocator, maximumWeight - item.weight, items[1..]) catch 0) + item.value;

    const nullValue = maximumValue(allocator, maximumWeight, items[1..]) catch 0;
    return @max(pickValue, nullValue);
}
